runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java

```diff
@@ -0,0 +1,330 @@
+package org.corfudb.runtime.object;
+
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.protocols.logprotocol.SMREntry;
+import org.corfudb.runtime.exceptions.NoRollbackException;
+import org.corfudb.runtime.object.transactions.AbstractTransactionalContext;
+import org.corfudb.runtime.view.StreamView;
+
+import java.util.*;
+import java.util.concurrent.locks.StampedLock;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+
+/**
+ * A version locked object. The version locked object contains a
+ * the state of an object as well as a history of modifications to
+ * the object, which can be optimistic.
+ * <p>
+ * Created by mwei on 11/13/16.
+ */
+@Slf4j
+public class VersionLockedObject<T> {
+
+    // maybe this shouldn't be constant.
+    static final int MAX_UNDO_SIZE = 50;
+
+    /**
+     * The actual underlying object.
+     */
+    T object;
+
+    /** The version of the underlying object.
+     *
+     */
+    long version;
+    long globalVersion;
+
+    /** A lock, which controls access to modifications to
+     * the object. Any access to unsafe methods should
+     * obtain the lock.
+     */
+    private StampedLock lock;
+
+    /** The stream view this object is backed by.
+     *
+     */
+    private StreamView sv;
+
+    /** If the object reflects optimistic updates, the
+     * context which made those updates.
+     */
+    private AbstractTransactionalContext modifyingContext;
+
+    /** An undo log, which records undo entries for the object.
+     *
+     */
+    private Deque<SMREntry> undoLog;
+
+    /** An optimistic undo log, which records undo entries for
+     * optimistic changes to the object.
+     */
+    private Deque<SMREntry> optimisticUndoLog;
+
+    /** True, if the object is optimistically modified.
+     *
+     */
+    private boolean optimisticallyModified;
+
+    /** True, if optimistic changes to this object can be undone.
+     *
+     */
+    private boolean optimisticallyUndoable;
+
+    /** The number of optimistic changes made to this object.
+     *
+     */
+    private int optimisticVersion;
+
+    /** The proxy which created this object.
+     *
+     */
+    private ICorfuSMRProxyInternal<T> proxy;
```

Comment:
This is a cycle. The versionlocked object contains an instance of internalProxy and the InternalProxy contains an instance of the versionLockedObject. Can we break this cycle ?